File: src/infrastructure/session/in_memory_discovery_adapter.py
"""In-memory implementation of DiscoveryPort.

Bridges the DiscoveryPort protocol to the DiscoverySession aggregate
using SessionStore for persistence between calls.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, cast

from src.domain.models.discovery_session import DiscoverySession

if TYPE_CHECKING:
    from src.domain.models.discovery_values import DiscoveryMode
    from src.domain.models.tech_stack import TechStack
    from src.infrastructure.session.session_store import SessionStore


class InMemoryDiscoveryAdapter:
    """In-memory adapter for guided discovery sessions.

    Wraps SessionStore and DiscoverySession aggregate, delegating
    all state transitions to the domain model.
    """

    def __init__(self, store: SessionStore) -> None:
        self._store = store

    def _get(self, session_id: str) -> DiscoverySession:
        """Retrieve and cast a session from the store."""
        return cast("DiscoverySession", self._store.get(session_id))

    def start_session(self, readme_content: str) -> DiscoverySession:
        """Start a new guided discovery session from README content.

        Args:
            readme_content: The raw text of the project README.

        Returns:
            The newly created DiscoverySession.
        """
        session = DiscoverySession(readme_content)
        self._store.put(session.session_id, session)
        return session

    def set_mode(self, session_id: str, mode: DiscoveryMode) -> DiscoverySession:
        """Set the discovery mode on a session.

        Args:
            session_id: The active discovery session identifier.
            mode: The DiscoveryMode to set.

        Returns:
            The updated DiscoverySession.
        """
        session = self._get(session_id)
        session.set_mode(mode)
        self._store.put(session_id, session)
        return session

    def set_tech_stack(self, session_id: str, tech_stack: TechStack) -> DiscoverySession:
        """Set the tech stack on a discovery session.

        Args:
            session_id: The active discovery session identifier.
            tech_stack: The TechStack value object to store.

        Returns:
            The updated DiscoverySession.
        """
        session = self._get(session_id)
        session.set_tech_stack(tech_stack)
        self._store.put(session_id, session)
        return session

    def detect_persona(self, session_id: str, choice: str) -> DiscoverySession:
        """Detect user persona for the given session.

        Args:
            session_id: The active discovery session identifier.
            choice: The persona choice ("1"-"4").

        Returns:
            The updated DiscoverySession.
        """
        session = self._get(session_id)
        session.detect_persona(choice)
        self._store.put(session_id, session)
        return session

    def answer_question(
        self, session_id: str, question_id: str, answer: str
    ) -> DiscoverySession:
        """Submit an answer to a discovery question.

        Args:
            session_id: The active discovery session identifier.
            question_id: The question being answered (e.g. "Q1").
            answer: The user's free-text answer.

        Returns:
            The updated DiscoverySession.
        """
        session = self._get(session_id)
        session.answer_question(question_id, answer)
        self._store.put(session_id, session)
        return session

    def skip_question(
        self, session_id: str, question_id: str, reason: str
    ) -> DiscoverySession:
        """Skip a question with an explicit reason.

        Args:
            session_id: The active discovery session identifier.
            question_id: The question to skip.
            reason: Why it was skipped (must be non-empty).

        Returns:
            The updated DiscoverySession.
        """
        session = self._get(session_id)
        session.skip_question(question_id, reason)
        self._store.put(session_id, session)
        return session

    def confirm_playback(
        self, session_id: str, confirmed: bool, corrections: str = ""
    ) -> DiscoverySession:
        """Confirm or reject a playback summary.

        Args:
            session_id: The active discovery session identifier.
            confirmed: True if the user confirms the playback.
            corrections: Optional correction text when rejecting playback.

        Returns:
            The updated DiscoverySession.
        """
        session = self._get(session_id)
        session.confirm_playback(confirmed, corrections)
        self._store.put(session_id, session)
        return session

    def get_session(self, session_id: str) -> DiscoverySession:
        """Retrieve a session by ID.

        Args:
            session_id: The session identifier to look up.

        Returns:
            The stored DiscoverySession.

        Raises:
            SessionNotFoundError: If no session matches the given ID.
        """
        return self._get(session_id)

    def complete(self, session_id: str) -> DiscoverySession:
        """Complete the discovery session and produce domain artifacts.

        Args:
            session_id: The active discovery session identifier.

        Returns:
            The completed DiscoverySession with events.
        """
        session = self._get(session_id)
        session.complete()
        self._store.put(session_id, session)
        return session

File: internal/discovery/infrastructure/in_memory_discovery_adapter.go
package infrastructure

import (
	"context"
	"fmt"

	discoverydomain "github.com/alty-cli/alty/internal/discovery/domain"
	domainerrors "github.com/alty-cli/alty/internal/shared/domain/errors"
	vo "github.com/alty-cli/alty/internal/shared/domain/valueobjects"
	"github.com/alty-cli/alty/internal/shared/infrastructure/persistence"
)

// InMemoryDiscoveryAdapter is an in-memory adapter for guided discovery sessions.
// Wraps SessionStore and DiscoverySession aggregate, delegating all state
// transitions to the domain model.
type InMemoryDiscoveryAdapter struct {
	store *persistence.SessionStore
}

// NewInMemoryDiscoveryAdapter creates a new InMemoryDiscoveryAdapter.
func NewInMemoryDiscoveryAdapter(store *persistence.SessionStore) *InMemoryDiscoveryAdapter {
	return &InMemoryDiscoveryAdapter{store: store}
}

func (a *InMemoryDiscoveryAdapter) getSession(sessionID string) (*discoverydomain.DiscoverySession, error) {
	val, err := a.store.Get(sessionID)
	if err != nil {
		return nil, fmt.Errorf("session '%s': %w", sessionID, domainerrors.ErrNotFound)
	}
	session, ok := val.(*discoverydomain.DiscoverySession)
	if !ok {
		return nil, fmt.Errorf("session '%s' has unexpected type: %w", sessionID, domainerrors.ErrNotFound)
	}
	return session, nil
}

// --- Discovery port methods (no context.Context) ---

// StartSession starts a new guided discovery session from README content.
func (a *InMemoryDiscoveryAdapter) StartSession(readmeContent string) (*discoverydomain.DiscoverySession, error) {
	session := discoverydomain.NewDiscoverySession(readmeContent)
	a.store.Put(session.SessionID(), session)
	return session, nil
}

// DetectPersona detects the user persona based on their choice.
func (a *InMemoryDiscoveryAdapter) DetectPersona(sessionID string, choice string) (*discoverydomain.DiscoverySession, error) {
	session, err := a.getSession(sessionID)
	if err != nil {
		return nil, err
	}
	if err := session.DetectPersona(choice); err != nil {
		return nil, err
	}
	a.store.Put(sessionID, session)
	return session, nil
}

// AnswerQuestion submits an answer to a discovery question.
func (a *InMemoryDiscoveryAdapter) AnswerQuestion(sessionID string, questionID string, answer string) (*discoverydomain.DiscoverySession, error) {
	session, err := a.getSession(sessionID)
	if err != nil {
		return nil, err
	}
	if err := session.AnswerQuestion(questionID, answer); err != nil {
		return nil, err
	}
	a.store.Put(sessionID, session)
	return session, nil
}

// SkipQuestion skips a question with an explicit reason.
func (a *InMemoryDiscoveryAdapter) SkipQuestion(sessionID string, questionID string, reason string) (*discoverydomain.DiscoverySession, error) {
	session, err := a.getSession(sessionID)
	if err != nil {
		return nil, err
	}
	if err := session.SkipQuestion(questionID, reason); err != nil {
		return nil, err
	}
	a.store.Put(sessionID, session)
	return session, nil
}

// ConfirmPlayback confirms or rejects the playback summary.
func (a *InMemoryDiscoveryAdapter) ConfirmPlayback(sessionID string, confirmed bool) (*discoverydomain.DiscoverySession, error) {
	session, err := a.getSession(sessionID)
	if err != nil {
		return nil, err
	}
	if err := session.ConfirmPlayback(confirmed, ""); err != nil {
		return nil, err
	}
	a.store.Put(sessionID, session)
	return session, nil
}

// Complete completes the discovery session and produces domain artifacts.
func (a *InMemoryDiscoveryAdapter) Complete(sessionID string) (*discoverydomain.DiscoverySession, error) {
	session, err := a.getSession(sessionID)
	if err != nil {
		return nil, err
	}
	if err := session.Complete(); err != nil {
		return nil, err
	}
	a.store.Put(sessionID, session)
	return session, nil
}

// --- Adapter-only methods (not part of Discovery port) ---

// GetSession retrieves a discovery session by ID.
func (a *InMemoryDiscoveryAdapter) GetSession(_ context.Context, sessionID string) (*discoverydomain.DiscoverySession, error) {
	return a.getSession(sessionID)
}

// SetTechStack sets the tech stack on a discovery session.
func (a *InMemoryDiscoveryAdapter) SetTechStack(_ context.Context, sessionID string, techStack vo.TechStack) (*discoverydomain.DiscoverySession, error) {
	session, err := a.getSession(sessionID)
	if err != nil {
		return nil, err
	}
	if err := session.SetTechStack(&techStack); err != nil {
		return nil, err
	}
	a.store.Put(sessionID, session)
	return session, nil
}

// SetMode sets the discovery mode on a session.
func (a *InMemoryDiscoveryAdapter) SetMode(_ context.Context, sessionID string, mode discoverydomain.DiscoveryMode) (*discoverydomain.DiscoverySession, error) {
	session, err := a.getSession(sessionID)
	if err != nil {
		return nil, err
	}
	if err := session.SetMode(mode); err != nil {
		return nil, err
	}
	a.store.Put(sessionID, session)
	return session, nil
}
